src/onediff/utils/lora.py

```diff
@@ -0,0 +1,438 @@
+    offload_weight="lora",
+) -> None:
+    assert isinstance(self, torch.nn.Conv2d)
+    if isinstance(self, DualModule):
+        self = self._torch_module
+    conv_unfuse_lora(self)
+    dtype, device = self.weight.data.dtype, self.weight.data.device
+
+    w_down = state_dict["lora.down.weight"].float().to(device)
+    w_up = state_dict["lora.up.weight"].float().to(device)
+
+    if alpha is not None:
+        w_up = w_up * alpha / rank
+
+    if offload_weight == "lora":
+        self.register_buffer("_lora_up", w_up.to(offload_device))
+        self.register_buffer(
+            "_lora_down", state_dict["lora.down.weight"].to(offload_device)
+        )
+        self._lora_scale = lora_scale
+    elif offload_weight == "weight":
+        self.register_buffer(
+            "_lora_orig_weight", self.weight.data.clone().to(offload_device)
+        )
+    else:
+        raise ValueError(f"Invalid offload weight: {offload_weight}")
+
+    lora_weight = torch.mm(w_up.flatten(start_dim=1), w_down.flatten(start_dim=1))
+    lora_weight = lora_weight.reshape((self.weight.shape)) * lora_scale
+
+    fused_weight = self.weight.data.float() + lora_weight
+    self.weight.data.copy_(fused_weight.to(device=device, dtype=dtype))
+
+
+def conv_unfuse_lora(self: torch.nn.Conv2d):
+    assert isinstance(self, torch.nn.Conv2d)
+
+    fused_weight = self.weight.data
+    dtype, device = fused_weight.dtype, fused_weight.device
+
+    if (
+        "_lora_orig_weight" in self._buffers
+        and self.get_buffer("_lora_orig_weight") is not None
+    ):
+        unfused_weight = self._lora_orig_weight
+        self._lora_orig_weight = None
+
+    elif "_lora_up" in self._buffers and self.get_buffer("_lora_up") is not None:
+        w_up = self._lora_up.to(device=device).float()
+        w_down = self._lora_down.to(device).float()
+
+        fusion = torch.mm(w_up.flatten(start_dim=1), w_down.flatten(start_dim=1))
+        fusion = fusion.reshape((fused_weight.shape))
+        unfused_weight = fused_weight.float() - (self._lora_scale * fusion)
+
+        self._lora_up = None
+        self._lora_down = None
+        self._lora_scale = None
+
+    else:
+        return
+
+    self.weight.data.copy_(unfused_weight.to(device=device, dtype=dtype))
+
+
+def load_and_fuse_lora(
+    pipeline: LoraLoaderMixin,
+    lora: Union[str, Path, Dict[str, torch.Tensor]],
+    lora_scale: float = 1.0,
+    adapter_name: Optional[str] = None,
+    *,
+    offload_device="cpu",
+    offload_weight="lora",
+    **kwargs,
+) -> None:
+    self = pipeline
+    if adapter_name is not None:
+        raise ValueError(f"load_and_fuse_lora is not supported for adapter_name")
+
+    state_dict, network_alphas = load_state_dict_cached(
+        lora, unet_config=self.unet.config, **kwargs
+    )
+
+    is_correct_format = all("lora" in key for key in state_dict.keys())
+    if not is_correct_format:
+        raise ValueError("Invalid LoRA checkpoint.")
+
+    # load lora into unet
```

Comment:
low_cpu_mem_usage = low_cpu_mem_usage if low_cpu_mem_usage is not None else _LOW_CPU_MEM_USAGE_DEFAULT